aws-crt/lib/aws-crt/errors.rb

```diff
@@ -6,14 +6,32 @@ class Error < StandardError
   module Errors
     @const_set_mutex = Mutex.new
 
+    AWS_TO_RUBY_ERROR_MAP = {
+      'AWS_ERROR_INVALID_INDEX' => IndexError,
+      'AWS_ERROR_OOM' => NoMemoryError,
+      'AWS_ERROR_UNIMPLEMENTED' => NotImplementedError,
+      'AWS_ERROR_INVALID_ARGUMENT' => ArgumentError,
+      'AWS_ERROR_SYS_CALL_FAILURE' => SystemCallError,
+      'AWS_ERROR_DIVIDE_BY_ZERO' => ZeroDivisionError,
+      'AWS_ERROR_HASHTBL_ITEM_NOT_FOUND' => KeyError
+    }.freeze
+
     def self.raise_last_error
       error_code = Aws::Crt::Native.last_error
-      error_name = Aws::Crt::Native.error_name(error_code)
-      raise error_class(error_name), Aws::Crt::Native.error_debug_str(error_code)
+      unless error_code.zero?
+        error_name = Aws::Crt::Native.error_name(error_code)
+        msg = Aws::Crt::Native.error_debug_str(error_code)
+        Aws::Crt::Native.reset_error
+        raise error_class(error_name), msg
+      end
     end
 
     # Get the error class for a given error_name
     def self.error_class(error_name)
+      if AWS_TO_RUBY_ERROR_MAP.include? error_name
+        return AWS_TO_RUBY_ERROR_MAP[error_name]
```

Comment:
what happens when this list grows? an error previously thrown will change

Reply:
Yeah - it would be a breaking change (add/remove from) this list.  We should review the errors here: https://github.com/awslabs/aws-c-common/blob/78dbfab579dea968a116a4e97f4af04b94e0d2b5/include/aws/common/error.h#L143
And see if there are any others we want to map.  This was what looked appropriate to me.

Reply:
consider instead always raising runtime errors with the common runtime error in the message

Reply:
Why? Many of these errors do map well to Ruby errors - eg OutOfMemory.  Mapping to specific exceptions (both Ruby standard exceptions where they make sense and otherwise to `SpecificErrorClass < Aws::Crt::Error < StandardExceptions` gives users the flexibility to catch and handle different errors as needed.

Reply:
it's a fragile design to have a breaking additive change. if we wrap every possible error into a single ruby error (i.e. Exception) with a variable message, then mapping can be done on top of that. scrap RuntimeError, since Exception is the base class, any rescues against Exception will still work if we wrap it with a new error in the future

Reply:
We do wrap all of the other errors under a single base exception (`Aws::Crt::Error` which extends `StandardError`).  However, there are some (a fairly static list) that I believe should be mapped to standard Ruby exceptions - eg out of memory error.  These will not change. 

I don't believe this is a fragile design here - this list of exception mappings will not change in a breaking way.  CRT may add new exceptions for new functionality and we may decide that those should map to existing ruby exceptions instead which is not breaking (new exception, new mapping).